contrib/cgi/src/cgi.c

```diff
@@ -0,0 +1,723 @@
+	}
+
+	reader = json_reader_new(json_parser_get_root(parser));
+
+	json_reader_read_member(reader, name);
+	value = json_reader_get_int_value(reader);
+	json_reader_end_member(reader);
+
+	g_object_unref(reader);
+	g_object_unref(parser);
+
+out:
+	return value;
+}
+
+/*
+ * The device will reboot shortly after the install process succeeded.
+ * This function determines if all requirements are met and the postinstall
+ * hook will fire soon.
+ */
+static gboolean get_will_reboot(RInstaller *installer, gboolean *will_reboot, GError **error)
+{
+	gboolean res = FALSE;
+	GError *ierror = NULL;
+	gboolean last_install_success = FALSE;
+	gchar* install_message = NULL;
+	gint install_percentage = -1;
+
+	g_return_val_if_fail(will_reboot != NULL, FALSE);
+
+	if (!get_last_installation_success(installer, &last_install_success, &ierror)) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	if (!get_progress_updated(installer, &install_percentage, &install_message, &ierror)) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	*will_reboot = last_install_success && install_percentage == 100 && g_file_test(STATUS_FILE_LOCATION, G_FILE_TEST_EXISTS);
+	res = TRUE;
+
+out:
+	g_free(install_message);
+
+	return res;
+}
+
+/* remove stale status and bundle files */
+static void try_cleanup_status_on_install_failure(RInstaller *installer)
+{
+	gboolean last_install_success = FALSE;
+	gchar* install_message = NULL;
+	gint install_percentage = -1;
+
+	if (!get_last_installation_success(installer, &last_install_success, NULL))
+		goto out;
+
+	if (!get_progress_updated(installer, &install_percentage, &install_message, NULL))
+		goto out;
+
+	if (!last_install_success && install_percentage == 100 && g_file_test(STATUS_FILE_LOCATION, G_FILE_TEST_EXISTS)) {
+		g_remove(STATUS_FILE_LOCATION);
+		g_remove(BUNDLE_TARGET_LOCATION);
+	}
+
+out:
+	g_free(install_message);
+}
+
+/*
+ * Write upload status information that is not available via D-Bus to status
+ * file.
+ */
+static gboolean write_upload_status(gint upload_progress, GError **error)
+{
+	gboolean res = FALSE;
+	JsonBuilder *builder = json_builder_new();
+	JsonGenerator *gen = NULL;
+	JsonNode *root = NULL;
+	const gchar **header_ptr;
+	gchar *header_name_upper = NULL;
+	gchar *full_header_name = NULL;
+	gchar *header_value = NULL;
+	GError *ierror = NULL;
+
+	json_builder_begin_object(builder);
+
+	/*
+	 * progression of the running upload operation in percent, any upload
+	 * not necessarily triggered by the client requesting the status value
+	 * from 0 to 100, -1 if no upload in progress
+	 */
+	json_builder_set_member_name(builder, "upload_progress");
+	json_builder_add_int_value(builder, upload_progress);
+
+	/*
+	 * value of a custom HTTP header that may be provided by the client
+	 * performing the upload used to identify who is currently uploading a
+	 * firmware
+	 */
+	header_ptr = &HTTP_HEADERS[0];
+	while (*header_ptr) {
+		header_name_upper = g_ascii_strup(*header_ptr, -1);
```

Comment:
Scope of `header_name_upper`, `full_header_name` and `header_value` variables can and should be limited to while loop.

Reply:
Right, will do.